File: app/src/main/java/com/ghostnexora/vpn/tunnel/Socks5OutboundProbe.java
package com.ghostnexora.vpn.tunnel;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.SNIHostName;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSocket;

/** Active loopback SOCKS -> Xray -> SSH -> TLS connectivity probe. */
final class Socks5OutboundProbe {
    private static final int CONNECT_TIMEOUT_MS = 4_000;
    private static final int IO_TIMEOUT_MS = 12_000;
    private static final InetAddress IPV4_LOOPBACK;

    static final List<SocksProbeTarget> targets = Collections.unmodifiableList(Arrays.asList(
            new SocksProbeTarget(new byte[]{1, 1, 1, 1}, 443, "one.one.one.one"),
            new SocksProbeTarget(new byte[]{8, 8, 8, 8}, 443, "dns.google")
    ));

    static {
        try {
            IPV4_LOOPBACK = InetAddress.getByAddress(new byte[]{127, 0, 0, 1});
        } catch (IOException impossible) {
            throw new ExceptionInInitializerError(impossible);
        }
    }

    private Socks5OutboundProbe() {
    }

    static SocksProbeResult measure(int localPort, SocksProbeTarget target) throws Exception {
        if (localPort < 1 || localPort > 65_535) {
            throw new IllegalArgumentException("Invalid Xray health-check port");
        }
        long startedAt = System.nanoTime();
        Socket socket = new Socket();
        try {
            socket.setTcpNoDelay(true);
            socket.setSoTimeout(IO_TIMEOUT_MS);
            socket.connect(new InetSocketAddress(IPV4_LOOPBACK, localPort), CONNECT_TIMEOUT_MS);

            InputStream input = socket.getInputStream();
            OutputStream output = socket.getOutputStream();
            negotiateNoAuthentication(input, output);
            requestIpv4Connect(input, output, target.getAddress(), target.getPort());

            SSLSocket tlsSocket = (SSLSocket) SSLContext.getDefault().getSocketFactory().createSocket(
                    socket,
                    target.getServerName(),
                    target.getPort(),
                    true
            );
            tlsSocket.setUseClientMode(true);
            tlsSocket.setSoTimeout(IO_TIMEOUT_MS);
            SSLParameters parameters = tlsSocket.getSSLParameters();
            parameters.setEndpointIdentificationAlgorithm("HTTPS");
            parameters.setServerNames(Collections.singletonList(new SNIHostName(target.getServerName())));
            tlsSocket.setSSLParameters(parameters);
            try (SSLSocket managed = tlsSocket) {
                managed.startHandshake();
                long latencyMs = Math.max(
                        1L,
                        TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedAt)
                );
                return new SocksProbeResult(
                        latencyMs,
                        managed.getSession().getProtocol() != null ? managed.getSession().getProtocol() : "",
                        managed.getSession().getCipherSuite() != null ? managed.getSession().getCipherSuite() : ""
                );
            }
        } finally {
            try {
                socket.close();
            } catch (Throwable ignored) {
            }
        }
    }

    static void negotiateNoAuthentication(InputStream input, OutputStream output) throws IOException {
        output.write(new byte[]{0x05, 0x01, 0x00});
        output.flush();

        byte[] response = new byte[2];
        readFully(input, response);
        if ((response[0] & 0xFF) != 0x05) {
            throw new IOException("Xray health-check returned an invalid SOCKS version");
        }
        if ((response[1] & 0xFF) != 0x00) {
            throw new IOException("Xray health-check rejected SOCKS no-authentication mode");
        }
    }

    static void requestIpv4Connect(
            InputStream input,
            OutputStream output,
            byte[] address,
            int port
    ) throws IOException {
        output.write(buildIpv4ConnectRequest(address, port));
        output.flush();
        readConnectReply(input);
    }

    static byte[] buildIpv4ConnectRequest(byte[] address, int port) {
        if (address == null || address.length != 4) {
            throw new IllegalArgumentException("SOCKS IPv4 address must contain four bytes");
        }
        if (port < 1 || port > 65_535) {
            throw new IllegalArgumentException("SOCKS target port is invalid");
        }
        return new byte[]{
                0x05,
                0x01,
                0x00,
                0x01,
                address[0],
                address[1],
                address[2],
                address[3],
                (byte) (port >>> 8),
                (byte) port
        };
    }

    static void readConnectReply(InputStream input) throws IOException {
        byte[] header = new byte[4];
        readFully(input, header);
        if ((header[0] & 0xFF) != 0x05) {
            throw new IOException("Xray health-check returned an invalid SOCKS reply");
        }
        int replyCode = header[1] & 0xFF;
        if (replyCode != 0x00) {
            throw new IOException("Xray SOCKS outbound rejected the probe (" + replyLabel(replyCode) + ")");
        }

        int addressLength;
        switch (header[3] & 0xFF) {
            case 0x01:
                addressLength = 4;
                break;
            case 0x03:
                addressLength = input.read();
                if (addressLength < 0) {
                    throw new IOException("Xray health-check returned a truncated domain reply");
                }
                break;
            case 0x04:
                addressLength = 16;
                break;
            default:
                throw new IOException("Xray health-check returned an invalid address type");
        }
        readFully(input, new byte[addressLength + 2]);
    }

    private static void readFully(InputStream input, byte[] destination) throws IOException {
        int offset = 0;
        while (offset < destination.length) {
            int read = input.read(destination, offset, destination.length - offset);
            if (read < 0) {
                throw new IOException("Xray health-check SOCKS connection closed early");
            }
            offset += read;
        }
    }

    private static String replyLabel(int code) {
        switch (code) {
            case 0x01: return "general failure";
            case 0x02: return "connection not allowed";
            case 0x03: return "network unreachable";
            case 0x04: return "host unreachable";
            case 0x05: return "connection refused";
            case 0x06: return "TTL expired";
            case 0x07: return "command not supported";
            case 0x08: return "address type not supported";
            default: return "code " + code;
        }
    }
}

final class SocksProbeResult {
    private final long latencyMs;
    private final String tlsProtocol;
    private final String cipherSuite;

    SocksProbeResult(long latencyMs, String tlsProtocol, String cipherSuite) {
        this.latencyMs = latencyMs;
        this.tlsProtocol = tlsProtocol;
        this.cipherSuite = cipherSuite;
    }

    public long getLatencyMs() { return latencyMs; }
    public String getTlsProtocol() { return tlsProtocol; }
    public String getCipherSuite() { return cipherSuite; }
}

final class SocksProbeTarget {
    private final byte[] address;
    private final int port;
    private final String serverName;

    SocksProbeTarget(byte[] address, int port, String serverName) {
        this.address = address.clone();
        this.port = port;
        this.serverName = serverName;
    }

    public byte[] getAddress() { return address.clone(); }
    public int getPort() { return port; }
    public String getServerName() { return serverName; }
    public String getEndpoint() { return "https://" + serverName + ":" + port; }
}

File: app/src/main/java/com/ghostnexora/vpn/tunnel/Socks5OutboundProbe.kt
package com.ghostnexora.vpn.tunnel

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import javax.net.ssl.SNIHostName
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSocket

/**
 * Verifies the active Xray outbound through its loopback-only SOCKS inbound.
 *
 * The probe performs this exact path:
 * local SOCKS client -> Xray -> SSH SOCKS bridge -> direct-tcpip -> TLS peer.
 * A successful TLS handshake proves bidirectional forwarding instead of only
 * proving that the native core process is running.
 */
internal object Socks5OutboundProbe {
    private const val CONNECT_TIMEOUT_MS = 4_000
    private const val IO_TIMEOUT_MS = 12_000
    private val IPV4_LOOPBACK = InetAddress.getByAddress(byteArrayOf(127, 0, 0, 1))
    val targets = listOf(
        SocksProbeTarget(byteArrayOf(1, 1, 1, 1), 443, "one.one.one.one"),
        SocksProbeTarget(byteArrayOf(8, 8, 8, 8), 443, "dns.google")
    )

    fun measure(localPort: Int, target: SocksProbeTarget): SocksProbeResult {
        require(localPort in 1..65535) { "Invalid Xray health-check port" }
        val startedAt = System.nanoTime()
        val socket = Socket()
        try {
            socket.tcpNoDelay = true
            socket.soTimeout = IO_TIMEOUT_MS
            socket.connect(InetSocketAddress(IPV4_LOOPBACK, localPort), CONNECT_TIMEOUT_MS)

            val input = socket.getInputStream()
            val output = socket.getOutputStream()
            negotiateNoAuthentication(input, output)
            requestIpv4Connect(input, output, target.address, target.port)

            val tlsSocket = SSLContext.getDefault().socketFactory.createSocket(
                socket,
                target.serverName,
                target.port,
                true
            ) as SSLSocket
            tlsSocket.useClientMode = true
            tlsSocket.soTimeout = IO_TIMEOUT_MS
            tlsSocket.sslParameters = tlsSocket.sslParameters.apply {
                endpointIdentificationAlgorithm = "HTTPS"
                serverNames = listOf(SNIHostName(target.serverName))
            }
            tlsSocket.use {
                it.startHandshake()
                val latencyMs = ((System.nanoTime() - startedAt) / 1_000_000L)
                    .coerceAtLeast(1L)
                return SocksProbeResult(
                    latencyMs = latencyMs,
                    tlsProtocol = it.session.protocol.orEmpty(),
                    cipherSuite = it.session.cipherSuite.orEmpty()
                )
            }
        } finally {
            runCatching { socket.close() }
        }
    }

    internal fun negotiateNoAuthentication(input: InputStream, output: OutputStream) {
        output.write(byteArrayOf(0x05, 0x01, 0x00))
        output.flush()

        val response = ByteArray(2)
        readFully(input, response)
        if ((response[0].toInt() and 0xFF) != 0x05) {
            throw IOException("Xray health-check returned an invalid SOCKS version")
        }
        if ((response[1].toInt() and 0xFF) != 0x00) {
            throw IOException("Xray health-check rejected SOCKS no-authentication mode")
        }
    }

    internal fun requestIpv4Connect(
        input: InputStream,
        output: OutputStream,
        address: ByteArray,
        port: Int
    ) {
        output.write(buildIpv4ConnectRequest(address, port))
        output.flush()
        readConnectReply(input)
    }

    internal fun buildIpv4ConnectRequest(address: ByteArray, port: Int): ByteArray {
        require(address.size == 4) { "SOCKS IPv4 address must contain four bytes" }
        require(port in 1..65535) { "SOCKS target port is invalid" }
        return byteArrayOf(
            0x05,
            0x01,
            0x00,
            0x01,
            address[0],
            address[1],
            address[2],
            address[3],
            (port ushr 8).toByte(),
            port.toByte()
        )
    }

    internal fun readConnectReply(input: InputStream) {
        val header = ByteArray(4)
        readFully(input, header)
        if ((header[0].toInt() and 0xFF) != 0x05) {
            throw IOException("Xray health-check returned an invalid SOCKS reply")
        }
        val replyCode = header[1].toInt() and 0xFF
        if (replyCode != 0x00) {
            throw IOException("Xray SOCKS outbound rejected the probe (${replyLabel(replyCode)})")
        }

        val addressLength = when (header[3].toInt() and 0xFF) {
            0x01 -> 4
            0x03 -> input.read().takeIf { it >= 0 }
                ?: throw IOException("Xray health-check returned a truncated domain reply")
            0x04 -> 16
            else -> throw IOException("Xray health-check returned an invalid address type")
        }
        readFully(input, ByteArray(addressLength + 2))
    }

    private fun readFully(input: InputStream, destination: ByteArray) {
        var offset = 0
        while (offset < destination.size) {
            val read = input.read(destination, offset, destination.size - offset)
            if (read < 0) throw IOException("Xray health-check SOCKS connection closed early")
            offset += read
        }
    }

    private fun replyLabel(code: Int): String = when (code) {
        0x01 -> "general failure"
        0x02 -> "connection not allowed"
        0x03 -> "network unreachable"
        0x04 -> "host unreachable"
        0x05 -> "connection refused"
        0x06 -> "TTL expired"
        0x07 -> "command not supported"
        0x08 -> "address type not supported"
        else -> "code $code"
    }
}

internal data class SocksProbeResult(
    val latencyMs: Long,
    val tlsProtocol: String,
    val cipherSuite: String
)

internal data class SocksProbeTarget(
    val address: ByteArray,
    val port: Int,
    val serverName: String
) {
    val endpoint: String
        get() = "https://$serverName:$port"
}
